### PR TITLE
[Test][Kernel] Tolerate one-ULP rounding in fused RMSNorm quant test

### LINKED ISSUE #53775: [Bug]: test_rms_norm fails deterministically on 26-SM MIG H200 CI agents (h200-ci-2-*), passes on 16-SM agents
## Summary

`tests/kernels/core/test_fused_quant_layernorm.py::test_rms_norm` fails **deterministically** on the subset of vLLM's H200 Buildkite agents whose MIG compute instance exposes **26 SMs** (`h200-ci-2-*`), and passes on the agents that expose **16 SMs** (`h200-ci-3-*`, `-4-*`, `-6-*`).

Because Buildkite assigns agents randomly per job, this presents as an intermittent, unexplained flake and repeatedly generates false "torch regression" signals in the torch version-bump lane. It is not a flake and it is not torch-version dependent.

Across 14 `Core Operation Kernels` jobs I inspected, the correlation is perfect: **every SM=26 job failed, every SM=16 job passed** — on both `main` and the torch-2.14 test lane.

## Evidence

| Lane | Build | Shard | Agent | MIG SM | Result |
|---|---|---|---|---|---|
| torch 2.14 | [84388](https://buildkite.com/vllm/ci/builds/84388) | 3 | `h200-ci-2-5` | **26** | **FAIL** |
| torch 2.14 | [84505](https://buildkite.com/vllm/ci/builds/84505#01a0175e-5a35-4617-8735-119b7c79fade) | 1 | `h200-ci-2-10` | **26** | **FAIL** |
| torch 2.14 | [84505](https://buildkite.com/vllm/ci/builds/84505#01a0175e-5a35-4133-9a03-65dbad12a26a) | 3 | `h200-ci-2-26` | **26** | **FAIL** |
| torch 2.14 | [85400](https://buildkite.com/vllm/ci/builds/85400#01a035ed-d784-4f96-bc1c-2e5a187152fa) | 3 | `h200-ci-2-15` | **26** | **FAIL** |
| **main** | **[85453](https://buildkite.com/vllm/ci/builds/85453#01a03782-6696-4c32-8927-425d1255f70b)** | **1** | **`h200-ci-2-13`** | **26** | **FAIL** |
| torch 2.14 | [84455](https://buildkite.com/vllm/ci/builds/84455) | 3 | `h200-ci-4-24` | 16 | pass |
| torch 2.14 | [84719](https://buildkite.com/vllm/ci/builds/84719) | 3 | `h200-ci-6-25` | 16 | pass |
| torch 2.14 | [84826](https://buildkite.com/vllm/ci/builds/84826) | 3 | `h200-ci-3-10` | 16 | pass |
| main | [85388](https://buildkite.com/vllm/ci/builds/85388) | 3 | `h200-ci-6-9` | 16 | pass |
| main | [85453](https://buildkite.com/vllm/ci/builds/85453) | 3 | `h200-ci-3-19` | 16 | pass |
| main | [85477](https://buildkite.com/vllm/ci/builds/85477) | 3 | `h200-ci-3-10` | 16 | pass |

(Plus three more SM=16 shard-1/shard-2 jobs that passed but do not contain these parametrizations.)

The `main` row is the important one: **build 85453, Shard 1, `h200-ci-2-13`** reproduces this on `main` with the standard `vllm-ci-postmerge-repo` image — i.e. with no torch bump involved at all.

## The failing parametrizations are stable per shard

Shard 3 (`h200-ci-2-*`) produces a byte-identical triple in builds 84388, 84505 and 85400:

```
test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-float8_e4m3fn-group-128-tma-0-scale-ub-False]
test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-float8_e4m3fn-per-token-tma-0-scale-ub-False]
test_rms_norm[False-cuda:0-0-dtype0-True-2048x1029-float8_e4m3fn-per-token-tma-0-scale-ub-False]
```

Shard 1 (`h200-ci-2-*`) produces its own byte-identical triple in 84505 **and in main build 85453**:

```
test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-int8-group-64-tma-0-scale-ub-False]
test_rms_norm[False-cuda:0-0-dtype0-True-2048x1024-float8_e4m3fn-group-64-tma-0-scale-ub-False]
test_rms_norm[True-cuda:0-0-dtype0-True-2048x1027-float8_e4m3fn-per-token-tma-0-scale-ub-False]
```

`dtype0` is `torch.bfloat16` (`DTYPES = [torch.bfloat16, torch.float]`), `add_residual=True`, `seed=0`, `scale_ub=False`, `tma_alignment=0` in every case. The failures span both the per-block (`group-64`, `group-128`) and per-token (`group_size=None`) paths, and both `int8` and `float8_e4m3fn`.

## The MIG geometry difference

Every agent reports the same GPU name, driver and CUDA version — the difference is only visible in the `MIG devices:` table's `SM` column:

```
h200-ci-2-15 (FAIL):
|  0    5   0   0  |              15MiB / 33280MiB    | 26      0 |  1   0    1    0    1 |

h200-ci-6-25 (pass):
|  0    3   0   0  |              15MiB / 33280MiB    | 16      0 |  1   0    1    0    1 |
```

Identical otherwise: `NVIDIA H200`, driver `580.159.03`, CUDA `13.0`, 33280 MiB slice, 29619 MiB BAR1, same CE/ENC/DEC/OFA/JPG counts.

A different SM count changes grid sizing and occupancy, which changes the reduction tree in the fused RMSNorm+quant kernel, which changes rounding. That is enough to move the fused-vs-reference comparison across the test's threshold.

## Why the CUDA path is fragile here

In `test_rms_norm`, ROCm has two bounded-outlier escapes that CUDA does not:

- `relax_block_rocm` (per-block bf16): relaxes the scale comparison to `rtol=1e-2` and tolerates a bounded count of fp8 ULP outliers — `max_outliers = ulp.numel() // 100_000 + 8`.
- `use_gfx950_fp8_allclose`: `fp8_allclose(rtol=0.125, atol=2e-3)` plus max ULP distance <= 1, with the comment *"Valid gfx950 reduction trees can straddle an E4M3 boundary."*

On CUDA both are `False`, so:

- **fp8** falls through to a whole-tensor `torch.allclose(a_deq, b_deq, rtol=5e-2, atol=5e-2)` with no outlier allowance.
- **int8** uses the strict `scales_close(rtol=1e-5, atol=1e-6)` — and the `rtol=1e-2` relaxation inside `scales_close` is also ROCm-only, so a single ULP flip in a group abs-max fails it.

The existing comment in the CUDA branch already anticipates exactly this situation:

> NOTE: It is possible that some future test cases trigger this max diff due to precision issues. If such an error is encountered, it's recommended to inspect the differences between all corresponding elements from each tensor ... (just a few bad elements should still be considered acceptable).

The reasoning behind the gfx950 escape — that a valid alternative reduction tree can straddle an E4M3 boundary — applies equally to a CUDA reduction tree that differs because the device exposes 26 SMs instead of 16.

## Reproduction

On a 26-SM MIG H200 partition:

```
pytest -v -s tests/kernels/core/test_fused_quant_layernorm.py::test_rms_norm \
  -k "float8_e4m3fn and per-token and 2048x1024 and not strided"
```

The same command on a 16-SM partition passes.

## Suggested directions

1. **Test robustness** — give the CUDA branch the same bounded-outlier treatment the ROCm branches already have, so a legitimate alternative reduction tree doesn't fail the assertion. At minimum, report *how many* elements exceed tolerance on failure; `assert ok` currently discards that, which is why this took several builds to characterise.
2. **CI fleet consistency** — the H200 pool is MIG-heterogeneous (26 SMs vs 16 SMs for the same 33 GB slice). Any tight-tolerance kernel numerics test is a coin flip on agent assignment. Normalising the MIG profile across the pool, or pinning numerics jobs to a single profile, would remove a whole class of phantom flakes.

Either alone would fix the symptom; (2) also protects every other numerics test in the suite.

## Impact

This has been misattributed to torch upgrades. I filed pytorch/pytorch#194740 against the torch-2.14 lane before finding the `main` reproduction on `h200-ci-2-13`; that issue is being closed as not-a-torch-bug. Anyone triaging the torch version-bump lane will keep hitting this until the agent geometry or the tolerance is addressed, because the "passes on main" baseline is only valid when the main job happened to land on a 16-SM agent.

## Caveats

Filed with the help of Claude Code, from automated triage of the vLLM CI logs. Specifically **not** verified by hand:

- The failure has not been reproduced interactively, and the element-level distribution of the tolerance breach is unknown (`assert ok` prints no diff).
- The SM-count -> reduction-tree -> rounding chain is a mechanism inferred from the code and the perfect 26-vs-16 correlation, not something I instrumented.
- Neither job log prints a torch version string; the torch difference between the `main` and 2.14 runs is inferred from the image repositories (`vllm-ci-postmerge-repo` vs `vllm-ci-test-repo`).

What *is* directly read from the logs: the agent hostnames, the MIG SM counts, the failing test IDs, and the pass/fail results in the table above.

### PR DESCRIPTION
**Pull request by nikhilkulkarni1755**
## Purpose

Fix #53775. `test_rms_norm` fails deterministically on the 26-SM H200 MIG agents and passes on the 16-SM ones, so the `Core Operation Kernels` job depends on agent assignment. Still true on `main` after #54111 (builds 87622, 87636, 87705, 87722).

`torch.randn` sizes its grid from `multiProcessorCount` (`calc_execution_policy`), so seed 0 draws a different tensor on each slice. Sweeping seeds 0–99 on an A10G: 132/3800 BF16 cases fail across 37 seeds. Every failure is a one-step rounding effect: the kernel and reference sum squares in different orders (CUB vs `torch.mean`), so `x * rms` can round to the adjacent BF16 value; when that element is a row/group abs-max the scale flips by one BF16 ULP and that row/group shifts one quantization step. The scale is also computed by different code on each side, which flips values on an fp8 tie even for fp32 inputs (15/1200 cases).

## Change

One rule replaces `relax_block_rocm` (#46658), `allow_cuda_fp8_rounding_outliers` (#54111) and the dequantized `rtol=5e-2` fallback, which cannot accept a 12.5% fp8 step:
- scales exact, or within `rtol=1e-2` for BF16 inputs;
- fp8 outputs within one ULP for at most `flipped_scales * row_or_group_len + numel // 100_000 + 8` elements;
- int8 outputs within 1, or within 2 for the same bounded count on BF16.

gfx950 branch unchanged. ROCm group BF16 fp8 outliers must now be within one ULP, not just bounded in count.

## Not a duplicate

No open PR references #53775 or touches this file (searched 2026-09-09).

## Test plan

A10G, `main` @ 385dce36b, `SEEDS` env-driven for the sweep only:
- baseline seeds 0–99, `-k "dtype0 and tma-0 and scale-ub-False and (2048x1024 or 2048x1027 or 2048x1029)"`: 132 failed / 3800
- those 132 IDs with this patch: 132 passed
- fp32 seeds 0–29 baseline: 15 failed / 1200; with patch: 15 passed
- full default file with patch: 672 passed

AI assistance (Claude Code) was used for CI log collection, sweep tooling and drafting; I reviewed every changed line and ran the tests above.

cc @mgoin @AndreasKaratzas

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
Not sure I understand this (eg what percentage of outliers does the proposed PR accept?). I would prefer if test changes are actually because of a CI failure.

**Comment by nikhilkulkarni1755** [comment]
Thanks for looking. This is driven by CI failures: `main` builds [87622](https://buildkite.com/vllm/ci/builds/87622), [87636](https://buildkite.com/vllm/ci/builds/87636), [87705](https://buildkite.com/vllm/ci/builds/87705), [87722](https://buildkite.com/vllm/ci/builds/87722) (2026-09-08) all fail `test_rms_norm` on the 26-SM `h200-ci-2-*` agents and pass on the 16-SM ones. That is #53775, still open after #54111. The seed sweep is only to show why: the same seed draws different inputs per SM count, and the strict CUDA check fails on 37/95 seeds on an A10G.

Accepted outliers, for 2048x1024 (2.1M elements): 28 baseline (0.0013%), plus one row (1024, 0.049%) or one group (64–128) per scale that is itself within one BF16 ULP. Every failing case in the sweep had exactly one flipped scale and at most 127 outliers. Happy to cap the number of flipped scales as well if you would prefer a fixed ceiling.
